astropy/io/fits/column.py

```diff
@@ -1528,7 +1528,16 @@ def _init_from_array(self, array):
         for idx in range(len(array.dtype)):
             cname = array.dtype.names[idx]
             ftype = array.dtype.fields[cname][0]
-            format = self._col_format_cls.from_recformat(ftype)
+
+            if ftype.kind == "O":
+                dtypes = {np.array(array[cname][i]).dtype for i in range(len(array))}
+                if len(dtypes) > 1:
+                    raise ValueError
```

Comment:
We should add an error message here.

Reply:
Yes probably better, currently it raises a `ValueError` so I used that but `TypeError` seems more appropriate in that case.